Codify deprecation process and improve documentation

scottisloud opened the pull request:
This PR introduces:
* Improvements to how we document deprecated deployments
* Provides people with a clearly stated process and timeline for deprecated deployments. 

The details
* Deprecated deployments will be moved to `/deprecated` 
* A notice will be posted to both README files in the root and `./deprecated`
* Date of deprecation and date of deletion will be included for each deployment method 
* Deletion will take place 12 weeks after deprecation.  
* Deployments already deprecated as of 2023-08-14 will be deleted in one month (2023-09-14), since both have been deprecated for > 8 months.

scottisloud commented on README.md:
```diff
@@ -26,20 +26,32 @@ After you've read the preparation guide, refer to the deployment example for you
-- [DigitalOcean App Platform](/deprecated/digitalocean-app-platform/)
+> 💡 **Note** that it is solely the _deployment method_ that is deprecated. Deprecating a deployment method is independent of the 1Password SCIM Bridge itself, or a specific version of the 1Password SCIM Bridge. For information about the latest 1Password SCIM Bridge version, please see the [changelog](https://app-updates.agilebits.com/product_history/SCIM).
+
+| Deployment                                                           | Deprecation Date | Deletion Date | Suggested Alternative                                                                                                              |
```
Currently this table exists in both README files. I don't love the duplication and it's a bit fiddly to maintain both. 

I'd like to maintain this list in one place. /deprecated makes sense insofar as we may want a list in that readme for navigations sake. 

But README at root is far more visible and ensures a deprecation isn't missed by people who may depend on it. 

Thoughts?

grellyd replied:
We don't need both.

I'd nix it from the Readme. Ensure there is a header on the deprecation documentation page, and link directly to it.

scottisloud commented on README.md:
```diff
@@ -26,20 +26,32 @@ After you've read the preparation guide, refer to the deployment example for you
+
+When a deployment method is deprecated, we will simultaneously post a notice to this README and move the corresponding example to [`/deprecated`](./deprecated/).
+
+Deprecated deployments will remain in [`/deprecated`](./deprecated/) for **three months**, after which time they will be deleted. The deletion date of deprecated deployments will be posted in [`/deprecated/README.md`](./deprecated/README.md).
```
Any thoughts on three months? I feel like that's plenty of time for people to clue in and make any required changes if they failed to make their own fork.

grellyd replied:
That sounds fine to me. 

I would personally add a weasel word in front of `three months`, such as about/around/approximately so we can sets dates that suit us, rather than exact periods of time. 

EG: if something gets deprecated on September 25th, let's not delete things on Christmas, but instead at the beginning of the new year.

grellyd commented on README.md:
```diff
@@ -26,20 +26,32 @@ After you've read the preparation guide, refer to the deployment example for you
+
+Deprecated deployments will remain in [`/deprecated`](./deprecated/) for **three months**, after which time they will be deleted. The deletion date of deprecated deployments will be posted in [`/deprecated/README.md`](./deprecated/README.md).
+
+Where possible, we will provide suggested alternatives in [`/deprecated/README.md`](./deprecated/README.md).
```
I'd suggest we also maintain a list of deleted deployment options. That means if you step away for a year, come back, and can't find your deployment, you'd at least be able to find it in the `No long supported` deployment method list.

grellyd commented on README.md:
```diff
@@ -26,20 +26,32 @@ After you've read the preparation guide, refer to the deployment example for you
+| [aws-ec2-terraform](./deprecated/aws-terraform/)                     | 2020-12-21       | 2023-09-14    | [AWS ECS Fargate with Terraform](./aws-ecsfargate-terraform/) or [AWS ECS Fargate with CloudFormation](./beta/aws-ecsfargate-cfn/) |
+| [DigitalOcean App Platform](./deprecated/digitalocean-app-platform/) | 2022-12-21       | 2023-09-14    | [Digital Ocean App Platform with `op` CLI](./beta/do-app-platform-op-cli/) or [Azure Container Apps](./beta/azure-container-apps/) |
+
+### Deprecation schedule
```
I'd move this section into the deprecation readme.

grellyd commented on deprecated/README.md:
```diff
@@ -1,8 +1,13 @@
+| --------------------------------------------------------- | ---------------- | ------------- | ------------------------------------------------------------------------------------------------------------------------------------ |
+| [aws-ec2-terraform](./aws-terraform)                      | 2020-12-21       | 2023-09-14    | [AWS ECS Fargate with Terraform](../aws-ecsfargate-terraform/) or [AWS ECS Fargate with CloudFormation](./beta/aws-ecsfargate-cfn/)  |
+| [DigitalOcean App Platform](./digitalocean-app-platform/) | 2022-12-21       | 2023-09-14    | [Digital Ocean App Platform with `op` CLI](../beta/do-app-platform-op-cli/) or [Azure Container Apps](../beta/azure-container-apps/) |
+
```
I'd add one other section to this document: guidance on how to deprecate a deployment for other Solution Architects to follow. 

EG:
1. Determine if we want to deprecate a deployment method 'baz' based on _ and _
1. Open a branch `deprecate/baz`
1. Move `baz` into `deprecated/baz`
1. Calculate a convenient deletion date approximately three months from now. A good day would be a workday without immediate holidays around it.
1. Add today's date to the `deployments` table
1. Add the deletion date to the `deployments` table
1. Put your MR up for review and approval

I'd also include exact instructions on how to do a deletion.

scottisloud replied:
Resolved in [92899a5](https://github.com/1Password/scim-examples/pull/249/commits/92899a5be260e7e08002ddd4dc1b1152648fefcc)

grellyd reviewed (changes requested):
A couple of changes we can make to clean it up a bit! I'd also suggest we document exact steps to ensure none of this process gets lost. Otherwise looks good.

scottisloud commented:
Initial attempt at addressing all feedback. Ready for formal review.

grellyd commented on deprecated/README.md:
```diff
@@ -1,8 +1,65 @@
+
+## Process for deprecating deployment methods
+
+Generally speaking we try to improve existing deployment methods or create additional deployment methods. However, where updating existing deployment methods is not feasible, we will resort to deprecation.
```
```suggestion
Generally speaking we try to improve existing deployment methods or create additional deployment methods. However, sometimes deployment methods become no longer relevant and so may be deprecated. 
```

grellyd commented on deprecated/README.md:
```diff
@@ -1,8 +1,65 @@
+All deprecations will take place through a merge request and must be approved by a 1Password employee.
+
+1. Identify a candidate for deprecation using the [above criteria](#process-for-deprecating-deployment-methods) along with insights from Integrations Support, your experience, and changes to dependencies, required utilities, host platforms, and technologies.
+   - There are no hard and fast rules here, and no single criteria. When making deprecation decisions, the focus should be on ensuring the best possible experience for 1Password SCIM Bridge users.
```
```suggestion
   - There are no hard and fast rules here, and no single criteria. When making deprecation decisions, the focus should be on ensuring the best possible experience for 1Password SCIM Bridge users. This does not immediately imply we maintain everything forever. To produce the best possible experience may require removing non-relevant deployment methods to enable us to focus on more relevant deployment methods. 
```

grellyd commented on deprecated/README.md:
```diff
@@ -1,8 +1,65 @@
+
+The following is a list of deployment methods that are no longer supported and were removed from the repository upon expiration of their deperecation period. These deployment methods are no longer supported. If you previously depended on one of these deployment methods, consider one of the suggested alternatives.
+
+| Deployment | Deprecation Date | Deletion Date | Suggested Alternative |
```
Do we want to perhaps link to the deletion MR?

scottisloud replied:
I had been turning this around in my head a bit myself. I think it's important we provide a rationale for transparency's sake, but we aren't likely going to do it justice in a table cell. 

I think your suggestion of a link to the MR for the `deprecate/<deployment-method>` branch in which the person who created the MR (as per my instructions) provides their rationale is the best approach. Prevents this readme from getting cluttered, but provides interested people recourse. 

I'll add a column to both tables for the MR link.

grellyd commented on deprecated/README.md:
```diff
@@ -1,8 +1,65 @@
+
+The following deployment methods are deprecated and will be removed from the repository on or around the Deletion Date.
+
+| Deployment                                                | Deprecation Date | Deletion Date | Suggested Alternative                                                                                                                |
```
Do we want to link to the deprecation MR? This would serve two purposes:
 1. A place for folks to go to add comments to discuss the deprecation
 1. A place for folks to go to find the old code, so that they can fork it/use it/export it as they need.

scottisloud replied:
I think that is a great idea. Likely this would be solved along with your comment [here](https://github.com/1Password/scim-examples/pull/249#discussion_r1296024161).

grellyd reviewed (dismissed):
A few minor comments and suggestions, but overall very shippable. ✅

plttn reviewed (approved):
lgtm

black-bryan reviewed (approved):
Great work on getting this documented. LGTM

scottisloud commented:
@plttn approved as I was about to share my summary of final changes. For the sake of completion prior to merging, in response to @grellyd's  feedback, I've:
- Added columns to both tables for links to the Deprecation (and deletion) PRs for future reference. This provides people who need access to that code an easy way to access it. It also provides people with an explanation for the deprecation if they want it, without cluttering up the README. 
- Updated deprecation and deletion instructions to include a directive to add PR links. 
- Added links to the PRs (such as they exist) for the currently-deprecated deployment methods. 
- Changed all references to "MR" to "PR" to align with GitHub's terminology.
